fix: For the openedx-events broker, use kafka _everywhere_ instead of redis

pwnage101 opened the pull request:
Devstack is currently split between services pre-configured to use redis, and others pre-configured to use kafka.  This results in the exceedingly dumb result of events produced via one broker, but never consumed because the consuming apps are hooked up to the other broker.

Just use the Kafka broker everywhere.  Why Kafka? This is more consistent with our 2U/edx.org deployment.

clintonb commented on README.rst:
```diff
@@ -16,21 +16,22 @@ from within a VM.
 
-This project was developed and tested using **Docker 1.13+**. If you are using
-macOS, please use `Docker for Mac`_. Previous Mac-based tools (e.g.
+This project was developed and tested using **Docker 1.13+**.
```
This statement can go away.

clintonb commented on README.rst:
```diff
@@ -16,21 +16,22 @@ from within a VM.
-macOS, please use `Docker for Mac`_. Previous Mac-based tools (e.g.
+This project was developed and tested using **Docker 1.13+**.
+
+For macOS users, please use `Docker for Mac`_ version 17.04+. Currently, 17.04
```
This is the minimum version for all OSes.

robrap replied:
@clintonb Are they using the same versioning scheme across all OSes?  Does that mean I should drop the line about:
```
This project was developed and tested using **Docker 1.13+**.
```
?

robrap replied:
Oops.  I missed the other comment.

clintonb replied:
Yes. Same version across all OSes.

robrap replied:
@clintonb I do need a little clarity here.  Are you saying that the versions are in sync across OSes in such that we will always recommend the same version bumps?  Or simply that 17.04 contains changes that we need for all OSes?

clintonb replied:
The Docker versions are in sync across all operating systems. If we are recommending 17.06 Edge for macOS, we are doing so for Windows and Linux, too.

robrap replied:
That keeps it simple enough. :)

jmbowman commented:
The other reason we currently need Docker edge is for building edxapp images, thanks to https://github.com/edx/configuration/pull/3864 .  The minimum required release for that one is 17.05.

robrap commented on README.rst:
```diff
@@ -16,21 +16,22 @@ from within a VM.
 
-This project was developed and tested using **Docker 1.13+**. If you are using
-macOS, please use `Docker for Mac`_. Previous Mac-based tools (e.g.
+This project requires **Docker 17.05+ CE**. Currently, 17.05 is only available
```
@jmbowman @clintonb: Even though we only need 17.05 for building images, I went for simple and just bumped the minimum to 17.05.

robrap commented on README.rst:
```diff
@@ -16,21 +16,22 @@ from within a VM.
+through the **Edge** version Docker 17.05 CE Edge.
+
+For macOS users, please use `Docker for Mac`_. Previous Mac-based tools (e.g.
 boot2docker) are not supported.
```
@clintonb I think this note is still relevant?

clintonb replied:
It is; although, will be less relevant in the future as folks forget about boot2docker.

robrap replied:
Thanks.  We'll leave it for now.

pwnage101 committed:
fix: For the openedx-events broker, use kafka _everywhere_ instead of redis

Devstack is currently split between services pre-configured to use
redis, and others pre-configured to use kafka.  This results in the
exceedingly dumb result of events produced via one broker, but never
consumed because the consuming apps are hooked up to the other broker.

brobro10000 commented on py_configuration_files/lms.py:
```diff
@@ -504,10 +507,11 @@ def should_show_debug_toolbar(request):  # lint-amnesty, pylint: disable=missing
 #################### Event bus backend ########################
-EVENT_BUS_PRODUCER = 'edx_event_bus_redis.create_producer'
-EVENT_BUS_REDIS_CONNECTION_URL = 'redis://:password@edx.devstack.redis:6379/'
+EVENT_BUS_KAFKA_SCHEMA_REGISTRY_URL = 'http://edx.devstack.schema-registry:8081'
+EVENT_BUS_KAFKA_BOOTSTRAP_SERVERS = 'edx.devstack.kafka:29092'
+EVENT_BUS_PRODUCER = 'edx_event_bus_kafka.create_producer'
+EVENT_BUS_CONSUMER = 'edx_event_bus_kafka.KafkaEventConsumer'
```
🥳

adamstankiewicz commented on docker-compose-host.yml:
```diff
@@ -71,6 +72,7 @@ services:
       - ${DEVSTACK_WORKSPACE}/src:/edx/src
       - ${PWD}/py_configuration_files/cms.py:/edx/app/edxapp/edx-platform/cms/envs/devstack.py
       - ${PWD}/py_configuration_files/lms.py:/edx/app/edxapp/edx-platform/lms/envs/devstack.py
+      - ${PWD}/py_private_requirements/cms.txt:/edx/private_requirements.txt
```
[curious] Are there no `private_requirements.txt` for the LMS?

For example, is there an equivalent like the following needed:

```
${PWD}/py_private_requirements/lms.txt:/edx/private_requirements.txt
```

pwnage101 replied:
Honestly, I don't know why settings are installed for both cms and lms for both images.  But, in this case for requirements it would be redundant because they go to the same destination path inside the container.

adamstankiewicz commented on docker-compose.yml:
```diff
@@ -449,7 +449,7 @@ services:
   lms:
     # Switch to `--settings devstack_with_worker` if you want to use lms-worker
-    command: bash -c 'source /edx/app/edxapp/edxapp_env && while true; do python /edx/app/edxapp/edx-platform/manage.py lms runserver 0.0.0.0:18000 --settings devstack; sleep 2; done'
+    command: bash -c 'source /edx/app/edxapp/edxapp_env && (pip install -r /edx/private_requirements.txt; while true; do python /edx/app/edxapp/edx-platform/manage.py lms runserver 0.0.0.0:18000 --settings devstack; sleep 2; done)'
```
Semi-related to my comment/question around LMS `private_requirements.txt` above; just trying to understand why these commands need to include `pip install -r /edx/private_requirements.txt` now when they didn't before.

pwnage101 replied:
Here are several loosely related facts which are required to understand the situation:

* openedx-events is architected to support either Redis or Kafka as the message broker, and the docs recommend to use the same in both your prod and dev environments.
* Redis python libraries are already installed in many/most of our django apps, including edx-platform.
* Kafka python libraries are NOT installed in openedx/edx-platform, and it seems like Axim wants to keep it that way.
* For the 2U edx.org prod environment, Kafka was chosen as the platform-wide message broker.  As such, we have custom python requirements [[1](https://github.com/edx/internal-dockerfiles/blob/master/edx-platform-private/Dockerfile#L33-L36), [2](https://github.com/edx/edx-internal/blob/fd6b33b9003a7b6ab191d8fffdb8d90aaeaa3419/argocd/applications/edxapp-lms/requirements/private_requirements.txt#L52-L53), [3](https://github.com/edx/edx-internal/blob/fd6b33b9003a7b6ab191d8fffdb8d90aaeaa3419/ansible/vars/edx.yml#L445-L446)] installed during our edxapp image build process to supplement the default requirements from `production.txt`.
* Devstack also used to be an openedx repo, and could not contain references to Kafka, therefore by default used the Redis broker. This is no longer a constraint because we forked Devstack.
* enterprise-subsidy in devstack was configured to use Kafka, and never successfully consumed events produced by edxapp (until this PR).

So, basically I needed a way to install extra Kafka python requirements into the edxapp container without modifying the edxapp image.  The images Devstack pulls are all the standard openedx images published to docker hub, and there's no pre-existing pipeline to build 2u-devstack-specific images, so adding a docker-compose step to install additional requirements seemed appropriate.

pwnage101 commented:
Updated PR to also switch `cms` and `course-discovery` apps to use Kafka as well.
